abi/messages.go

```diff
@@ -1,984 +1,331 @@
-		var resStonfiPaymentRequestMsgBody StonfiPaymentRequestMsgBody
-		err = tlb.Unmarshal(cell, &resStonfiPaymentRequestMsgBody)
-		if err == nil && !completedRead(cell) {
-			err = ErrStructSizeMismatch
-		}
-		if err == nil {
-			return StonfiPaymentRequestMsgOp, resStonfiPaymentRequestMsgBody, nil
+		return nil, nil, nil, err
+	}
+}
+
+func decodeMultipleMsgs(funcs []msgDecoderFunc, tag string) msgDecoderFunc {
+	return func(cell *boc.Cell) (*uint32, *MsgOpName, any, error) {
+		for _, f := range funcs {
+			tag, opName, object, err := f(cell)
+			if err == nil && completedRead(cell) {
+				return tag, opName, object, err
+			}
+			cell.ResetCounters()
+		}
+		return nil, nil, nil, fmt.Errorf("no one message can be unmarshled for %v", tag)
+	}
+}
+
+type InMsgBody struct {
+	SumType MsgOpName
+	OpCode  *uint32
+	Value   any
+}
+
+func (body *InMsgBody) UnmarshalTLB(cell *boc.Cell, decoder *tlb.Decoder) error {
+	body.SumType = UnknownMsgOp
+	tag, name, value, err := InternalMessageDecoder(cell, nil)
+	if err != nil {
+		return err
+	}
+	if tag == nil && name == nil {
+		body.SumType = EmptyMsgOp
+		return nil
+	}
+	body.OpCode = tag
+	if name != nil {
+		body.SumType = *name
+		body.Value = value
+	} else {
+		body.Value = cell
+	}
+	return nil
+}
+
+func (body *InMsgBody) UnmarshalJSON(data []byte) error {
+	var r struct {
+		SumType string
+		OpCode  *uint32
+		Value   json.RawMessage
+	}
+	if err := json.Unmarshal(data, &r); err != nil {
+		return err
+	}
+	body.SumType = r.SumType
+	body.OpCode = r.OpCode
+	if body.SumType == EmptyMsgOp {
+		return nil
+	}
+	if body.SumType == UnknownMsgOp {
+		c := boc.NewCell()
+		err := json.Unmarshal(r.Value, c)
+		if err != nil {
+			return err
+		}
+		body.Value = c
+		return nil
+	}
+	t, ok := KnownMsgInTypes[body.SumType]
+	if !ok {
+		return fmt.Errorf("unknown message body type '%v'", body.SumType)
+	}
+	o := reflect.New(reflect.TypeOf(t))
+	err := json.Unmarshal(r.Value, o.Interface())
+	if err != nil {
+		return err
+	}
+	body.Value = o.Elem().Interface()
+	return nil
+}
+
+func (body InMsgBody) MarshalJSON() ([]byte, error) {
+	if body.SumType == EmptyMsgOp {
+		return []byte("{}"), nil
+	}
+	buf := bytes.NewBufferString(`{"SumType": "` + body.SumType + `",`)
+	if body.OpCode != nil {
+		fmt.Fprintf(buf, `"OpCode":%v,`, *body.OpCode)
+	}
+	buf.WriteString(`"Value":`)
+	if body.SumType == UnknownMsgOp {
+		c, ok := body.Value.(*boc.Cell)
+		if !ok {
+			return nil, fmt.Errorf("unknown MsgBody should be Cell")
+		}
+		b, err := c.ToBoc()
+		if err != nil {
+			return nil, err
+		}
+		buf.WriteRune('"')
+		hex.NewEncoder(buf).Write(b)
+		buf.WriteString(`"}`)
+		return buf.Bytes(), nil
+	}
+	if KnownMsgInTypes[body.SumType] == nil {
+		return nil, fmt.Errorf("unknown MsgBosy type %v", body.SumType)
+	}
+	b, err := json.Marshal(body.Value)
+	if err != nil {
+		return nil, err
+	}
+	buf.Write(b)
+	buf.WriteRune('}')
+	return buf.Bytes(), nil
+}
+
+type ExtOutMsgBody struct {
```

Comment:
Both types `InMsgBody` and `ExtOutMsgBody` look similar and it would be great to simplify both by refactoring common code into a set of functions: 
```golang

func unmarshalMsgTLB(cell *boc.Cell, decoder *tlb.Decoder) (MsgOpName, *uint32, any, error) {
	tag, name, value, err := InternalMessageDecoder(cell, nil)
	if err != nil {
		return "", nil, nil, err
	}
	if tag == nil && name == nil {
		return EmptyMsgOp, nil, nil, nil
	}
	if name != nil {
		return *name, tag, value, nil
	}
	return UnknownMsgOp, tag, cell, nil
}

func (body *InMsgBody) UnmarshalTLB(cell *boc.Cell, decoder *tlb.Decoder) error {
	sumType, opCode, value, err := unmarshalMsgTLB(cell, decoder)
	if err != nil {
		return err
	}
	body.SumType = sumType
	body.OpCode = opCode
	body.Value = value
	return nil
}

func (b *ExtOutMsgBody) UnmarshalTLB(cell *boc.Cell, decoder *tlb.Decoder) error {
	sumType, opCode, value, err := unmarshalMsgTLB(cell, decoder)
	if err != nil {
		return err
	}
	b.SumType = sumType
	b.OpCode = opCode
	b.Value = value
	return nil
}
```

MarshalJSON and UnmarshalJSON could be simplified a lot